org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/InstanceOfExpression.java

```diff
@@ -299,20 +244,38 @@ public TypeBinding resolveType(BlockScope scope) {
 		this.pattern.setExpressionContext(ExpressionContext.INSTANCEOF_CONTEXT);
 		this.pattern.setExpectedType(this.expression.resolvedType);
 		this.pattern.resolveType(scope);
+
+		switch (this.expression.bits & ASTNode.RestrictiveFlagMASK) {
+
+			case Binding.FIELD:
+			case Binding.LOCAL:
```

Comment:
Here is my only concern - as discussed - shouldn't we use addAssignment for Field? Local is fine, but for field, in the light of a possible data race, won't the same thread have chances of seeing different values at different points if the evaluation happens multiple times? an addAssignment will take care of this. Also agree that there indeed is an inherent problem of not protecting/synchronizing in the first place - but again different values in the same thread for a "local" variable should be discussed further. thanks!

Reply:
Thanks for the comment - Manoj. I'll go ahead and make the requested change before merging. I see that the reference compiler uses hidden local when the expression is a field - may be you are onto something.